### DISCUSSION TITLE
Tree View in react-aria

### DISCUSSION THREAD

**Discussion by bstro** [adobe/react-spectrum#3551] [Q&A] [answered]
I need a [TreeView](https://spectrum.adobe.com/page/tree-view/) for my app and I'd like to use react-aria to build it out.  I'm wondering if there's any prior art in react-spectrum/react-aria? I haven't found anything so far. I'm wondering what tools in react-aria/stately would be most suitable for putting something like this together?

Off the top of my head, I was thinking:
* for state: [useTreeState](https://react-spectrum.adobe.com/react-stately/useTreeState.html) or [useTreeData](https://react-spectrum.adobe.com/react-stately/useTreeData.html) (it isn't clear to me what the difference is yet) — I'm not sure if these are going to be an exact fit given that the tree is recursive, but maybe they can be of some use.
* for behavior: [useGridList](https://react-spectrum.adobe.com/react-aria/useGridList.html)? Since the list items in a GridList [can include interactive elements](https://react-spectrum.adobe.com/react-aria/useGridList.html#features), maybe one of those interactive elements can be another instance of a GridList?

If I can't use a pre-made hook like `useGridList`, then I guess I would reach for [FocusScope](https://react-spectrum.adobe.com/react-aria/FocusScope.html) to manage focus, and `useKeyboard` for selecting individual items within the list (and the right-arrow key for drilling into a collapsed parent item).

I would love to hear if there is anything I'm missing? Thank you!

**Comment by devongovett** [comment, marked as answer]
We have plans to implement a tree view component in React Aria eventually. See #2396 for how we plan to implement it.

GridList is not quite right because it doesn't have descendants. The aria `treegrid` role is the right one to use.

**Reply by bstro** [reply]
This is _exactly_ what I was looking for. Thank you so much.
